File: code-r/solutions/07_read_solutions.R
# # Reading and Writing Data

# Copyright © 2010–2018 Cloudera. All rights reserved.
# Not to be reproduced or shared without prior written 
# consent from Cloudera.


# Sample answers.  You can run this standalone, or execute selected statements 
# in a wider session.


# ## Setup

library(sparklyr)

# Connect local session
config <- spark_config()
config$spark.driver.host <- Sys.getenv("CDSW_IP_ADDRESS")
spark <- spark_connect(
  master = "local",
  app_name = "read_solutions",
  config = config
)

# Reset `data` directory in HDFS
system("hdfs dfs -rm -r -skipTrash data")
system("hdfs dfs -mkdir data")


# ## Exercises

# (1) Read the raw drivers file into a Spark DataFrame.

# View a few records to verify the file format (could also be done in Hue)
system("hdfs dfs -cat /duocar/raw/drivers/* | head -3")

# Comma-delimited, header present.  Read the csv file.
drivers <- spark_read_csv(
  sc = spark,
  name = "drivers",
  path = "/duocar/raw/drivers/",
  header = TRUE,
  infer_schema = TRUE,
  delimiter = ",")

# Verify
drivers


# (2) Save the drivers DataFrame as a JSON file in your `data` directory.

# (guessing a syntax similar to the one for parquet)
spark_write_json(
  drivers,
  path = "data/drivers_json")

# Verify (could also be viewed in Hue)
system("hdfs dfs -cat data/drivers_json/* | head -3")


# (3) Read the drivers JSON file into a Spark DataFrame.

# (guessing syntax similar to the one for parquet)
drivers_json <- spark_read_json(
  sc = spark,
  name = "drivers_json",
  path = "data/drivers_json")


# (4) Delete the JSON file.

# Cleanup directory from HDFS
system("hdfs dfs -rm -r -skipTrash data/drivers_json")


# ## Cleanup

# Stop the `SparkSession`
spark_disconnect(spark)

File: code-python/solutions/07_read_solutions.py
# # Reading and Writing Data - Solutions

# Copyright © 2010–2018 Cloudera. All rights reserved.
# Not to be reproduced or shared without prior written 
# consent from Cloudera.


# ## Setup

# Create a SparkSession:
from pyspark.sql import SparkSession
spark = SparkSession.builder.master("local").appName("read_solutions").getOrCreate() 

# Read the raw rider data from HDFS:
riders = spark.read.csv("/duocar/raw/riders/", header=True, inferSchema=True)


# ## Exercises

# (1) Use the `json` method of the `DataFrameWriter` class to write the
# `riders` DataFrame to the `data/riders_json/` (HDFS) directory.

riders.write.json("data/riders_json/", mode="overwrite")

# (2) Use the `hdfs dfs -ls` command to list the contents of the
# `data/riders_json/` directory.

!hdfs dfs -ls data/riders_json

# (3) Use the `hdfs dfs -cat` and `head` commands to display a JSON file in
# the `data/riders_json` directory.

!hdfs dfs -cat data/riders_json/part* | head -n 5

# (4) Use Hue to browse the `data/riders_json/` directory.

# Note that the JSON file is noticeably larger than the original CSV file.

# (5) Use the `json` method of the `DataFrameReader` class to read the JSON
# file into a DataFrame.

riders_json = spark.read.json("data/riders_json/")

# (6) Examine the schema of the DataFrame.  Do you notice anything different?

riders_json.printSchema()

# Note that the columns are in alphabetical order.


# ## Cleanup

# Stop the SparkSession:
spark.stop()
